packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -4982,15 +4982,15 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
   /// ```
   /// {@end-tool}
   @optionalTypeArgs
-  void pop<T extends Object?>([ T? result ]) {
+  Future<void> pop<T extends Object?>([ T? result ]) async {
     assert(!_debugLocked);
     assert(() {
       _debugLocked = true;
       return true;
     }());
     final _RouteEntry entry = _history.lastWhere(_RouteEntry.isPresentPredicate);
     if (entry.pageBased) {
-      if (widget.onPopPage!(entry.route, result) && entry.currentState == _RouteLifecycle.idle) {
+      if (await widget.onPopPage!(entry.route, result) && entry.currentState == _RouteLifecycle.idle) {
```

Comment:
In the time we are awaiting this, we are `_debugLocked`, so if other navigation events occur while we wait people will see random crashes, that will be hard to debug and reproduce. Wondering if this API creates a footgun.

Reply:
I can think of several ways.

1. Use snychronously future if the widget.onPopPage returns value immediately.
2. refactor all imperative APIs so that they will wait for pop to finish before execute

I think 1. would make sure the existing app works as before.

2 seems like a real fix to the problem, but I am not sure if that will create some other issues.

WDYT?

Reply:
Talk to @goderbauer offline. I think we don't need to make this async to support our use case.

Here is what we should do

1. when onPopPage is called, go_router always veto it.
2. go_router calls the onExit callback and await the result
3. if the result come back yes, pop the goRoute in RouteMatchList

Reply:
Ok so I can close this PR, right?

Reply:
I will close this pr